### ISSUE TITLE
False positive when checking a string against a list/set of literals

### ISSUE THREAD

**Issue by msrd0** [microsoft/pyright#11701]
**Describe the bug**

When checking a string against a set or list of literals, pyright does not support using the string itself as a literal type afterwards. In the following example, the first `if` works and does not trigger any errors, while the second one (which I believe does semantically the same) triggers a type error:

```python
from typing import Literal


def foo(s: str) -> None:
	if s == "bar" or s == "foo":
		bar(s)
	if s in {"bar", "foo"}:  # Same behaviour if using ["bar", "foo"] instead
		bar(s)

def bar(s: Literal["bar", "foo"]) -> None:
	print(s)
```

```
/tmp/test.py
  /tmp/test.py:8:7 - error: Argument of type "str" cannot be assigned to parameter "s" of type "Literal['bar', 'foo']" in function "bar"
    Type "str" is not assignable to type "Literal['bar', 'foo']"
      "str" is not assignable to type "Literal['bar']"
      "str" is not assignable to type "Literal['foo']" (reportArgumentType)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**

Command Line and [Gram](https://codeberg.org/GramEditor/Gram) on Linux

```
$ pyright --version
pyright 1.1.412
```

**Comment by bschnurr** [comment]
Thanks for the report. This is working as designed and is the same behavior discussed in https://github.com/microsoft/pyright/issues/11166.

Pyright retains the literal types for a tuple expression like ("bar", "foo"), so membership testing can narrow s. List and set expressions are inferred as list[str] and set[str], respectively, so their individual literal values are not available for narrowing.

You can use a tuple here:

if s in ("bar", "foo"):
    bar(s)
